### PR TITLE
Carefully enhance file and file_base modules

### PR DESCRIPTION
**Pull request by frispete**
to support file encryption in a extensible way.

The changes are backwards compatible, hence no harm is done for any existing users.

For the next step, a decision has to be made: move file_base to keyring.backend or not?
@jaraco, what do you think?

### CODE REVIEW

File: tests/test_crypto.py [hunk 1]
```diff
@@ -1,5 +1,8 @@
 import unittest
-from unittest import mock
+try:
```
Comment (jaraco) [inline tests/test_crypto.py]: This change is unnecessary as long as backports.unittest_mock is installed, which it should be by tox and tests/requirements.

File: keyrings/alt/file.py [hunk 2]
```diff
@@ -37,7 +39,8 @@ class Encrypted(object):
     """
     PyCrypto-backed Encryption support
     """
-
+    scheme = 'PyCrypto [PBKDF2] AES256.CFB'
```
Comment (jaraco) [inline keyrings/alt/file.py]: Might be nice if the scheme were implementation agnostic - such that it only specifies the algorithm, and implementations could be swapped out. Is there a standard for doing that? When I encrypt something using https://www.infoencrypt.com/, it creates a header indicating the algorithm. Perhaps that's a model to follow. Just an idea for consideration, not a blocker for this change.

**Comment by jaraco** [comment]
> move file_base to keyring.backend or not?

I'm on the fence. Here's my hesitation - I don't want to encourage file-based backends by supporting them in the base library. On the other hand, the options remaining are (b) create a new library just for this base functionality (probably overkill) or (c) require that all file-based backends rely on keyrings.alt (which brings its own baggage). I'm happy to go with any of the three directions. What's your preference as the implementer of the first consuming lib? Let's go with your instinct.

File: tests/test_crypto.py (hunk 1 above)
Comment (frispete) [inline tests/test_crypto.py]: Can you drop that patch, or should I recreate the branch?

File: keyrings/alt/file.py (hunk 2 above)
Comment (frispete) [inline keyrings/alt/file.py]: Yeah, that would be great, but as of now, there're too many diverging crypto projects.
As always, the devil is in the details, and covering even the major ones is, ahem, ambitious...

**Comment by frispete** [comment]
My gut feeling is, that keeping them in keyrings.alt is the best option for now. Sure, this package carries some cruft, but as long as it builds, I'm fine with it.

**Comment by jaraco** [comment]
Sounds good. So is this PR ready? Could you do one more thing and add an entry to the CHANGES.rst file explaining the change?

**Comment by frispete** [comment]
Done.

**Comment by frispete** [comment]
With these patchsets, file.py and file_base.py are in sync with my project.

Major change is preparation for handling associated data, which is the result of discussing my approach at http://crypto.stackexchange.com. Since the built-in CFB mode is conceptually outdated, I refrained from adding some kind of associated data band-aid, which might give people the impression of being a secure scheme.

In [keyrings.cryptfile](https://github.com/frispete/keyrings.cryptfile.git), this infrastructure is used correctly. BTW, it gained support for all major AES AEAD modes, that PyCryptodome supports: CCM, EAX, GCM and OCB. 

While at it, raise related test coverage significantly..

I've prepared this changeset by cherry picking the relevant data from my project. Hopefully, all is well.

File: keyrings/alt/file.py [hunk 3]
```diff
@@ -78,9 +81,9 @@ def priority(self):
             __import__('Crypto.Cipher.AES')
             __import__('Crypto.Protocol.KDF')
             __import__('Crypto.Random')
-        except ImportError:
+        except ImportError:     # pragma: no cover
```
Comment (jaraco) [inline keyrings/alt/file.py]: I'm not sure I understand the purpose of these pragmas. They feel like noise in the code. Plus, this code doesn't seem like code that we don't expect to be executed. It seems like code that's legitimately not covered by the tests.

File: keyrings/alt/file.py [hunk 4]
```diff
@@ -1,34 +1,36 @@
...
+import getpass
 
 from keyring.py27compat import configparser
 
 from keyring.util import properties
 from keyring.util.escape import escape as escape_for_ini
 
-from . import file_base
-
+from keyrings.alt.file_base import (
```
Comment (jaraco) [inline keyrings/alt/file.py]: I prefer the relative imports. It allows for shorter lines and easier migrations of code. Plus, it's less repetition. Plus, why not take advantage of the clearer context of the module name?

```
from . import file_base
```

Then use `file_base.Keyring` and `file_base.decodebytes` below. It'll be a nicer looking import section plus code below that's easier to recognize the context (without having to read the import section). Moreover, it gives the advantage of being able to leverage the namespace, so if you have code here that presents the same concept, it can re-use the name (such as where [setuptools overrides sdist](https://github.com/pypa/setuptools/blob/master/setuptools/command/sdist.py#L24)).

Comment (jaraco) [inline keyrings/alt/file.py]: If it turns out that `file_base.decodebytes` is too wordy below, then maybe add an additional import for those two functions:

```
from .filebase import decodebytes, encodebytes
```

File: keyrings/alt/file_base.py [hunk 5]
```diff
@@ -43,16 +75,20 @@ class Keyring(FileBacked, KeyringBackend):
     """
 
     @abc.abstractmethod
-    def encrypt(self, password):
+    def encrypt(self, password, assoc = None):
         """
         Given a password (byte string), return an encrypted byte string.
+
+        assoc might provide associated data (typically: service and username)
```
Comment (jaraco) [inline keyrings/alt/file_base.py]: The docstring should describe what assoc is (byte string) and what it does (what's its purpose).

File: keyrings/alt/file_base.py [hunk 6]
```diff
@@ -61,6 +97,7 @@ def get_password(self, service, username):
         """
         service = escape_for_ini(service)
         username = escape_for_ini(username)
+        assoc = (service + '\0' + username).encode()
```
Comment (jaraco) [inline keyrings/alt/file_base.py]: Does it make sense for assoc to be a bytestring? Could it be a dict or tuple or just text?

The null byte here is disconcerting. It feels like a `C-style` hack. Why the null byte and not some other arbitrary character?

File: keyrings/alt/file_base.py [hunk 7]
```diff
@@ -71,35 +108,44 @@ def get_password(self, service, username):
...
             password = None
         return password
 
     def set_password(self, service, username, password):
         """Write the password in the file.
         """
-        service = escape_for_ini(service)
-        username = escape_for_ini(username)
-
+        assoc = (escape_for_ini(service) + '\0' +
```
Comment (jaraco) [inline keyrings/alt/file_base.py]: This looks similar, though incongruent to the code in get_password. Perhaps this behavior should be consolidated into a private method to capture the commonality and explain the purpose of the null byte.

**Comment by frispete** [comment]
Hi @jaraco, thank you for merging.

As for your valuable comments, they are much appreciated.

I'm about to leave, will reply to and fix them soon.

File: keyrings/alt/file.py (hunk 4 above)
Comment (frispete) [inline keyrings/alt/file.py]: You're right, of course. Will fix (..was a fallout of forking the module into another package and cherry pick the changes)

File: keyrings/alt/file.py (hunk 3 above)
Comment (frispete) [inline keyrings/alt/file.py]: Hmm, any idea of how to test these exceptions, e.g. how to raise these ImportError exceptions, while Crypto _is_ installed? Is that mockable?

File: keyrings/alt/file_base.py (hunk 5 above)
Comment (frispete) [inline keyrings/alt/file_base.py]: Will fix, patch is coming.

File: keyrings/alt/file_base.py (hunk 6 above)
Comment (frispete) [inline keyrings/alt/file_base.py]: Because I wanted to be sure to get exactly the service and username bytestrings (okay, due to value escaping, this is tamper resistant already, hmm..) 

assoc is a sensible value, that needs to be reproduced identically during decoding (or resulting in MAC failures). When using a dict, it will have to be an ordered one, but I avoided to act too subtle in this area. A tuple would work, but that has to be serialized to a byte string anyway.

File: keyrings/alt/file_base.py (hunk 7 above)
Comment (frispete) [inline keyrings/alt/file_base.py]: Yes, the divergence is due to avoidance of repeated escapes of the same value in get_password(), that would happen, if consolidated (probably a case of premature optimization).

We reuse the escaped values in get_password, while preserving the unescaped values in set_password (as escaping is delayed to _write_config_values in this code path).

Comment (frispete) [inline keyrings/alt/file_base.py]: I can consolidate this in a method _generate_assoc, but would like to keep the null byte, since I'm using these keyrings in other projects already... Blame it on my age :wink:

File: keyrings/alt/file.py (hunk 3 above)
Comment (jaraco) [inline keyrings/alt/file.py]: Looks like you can [use mock on sys.modules](http://stackoverflow.com/a/31718136/70170) if you want to make sure it's covered.
